### PR TITLE
 zeroize support 

### PR DESCRIPTION
**Pull request by charleschege**
Solve bugs by implementing zeroize manually

### CODE REVIEW

File: tai64/Cargo.toml [hunk 1]
```diff
@@ -17,10 +17,12 @@ maintenance = { status = "passively-maintained" }
 [dependencies]
 chrono = { version = "0.4", optional = true, default-features = false }
 serde = { version = "1", optional = true, default-features = false }
+zeroize = { version = "1.1.0", optional = true, default-features = false, features = ["zeroize_derive"] }
 
 [features]
 default = ["std"]
 std = []
+zero_out = ["zeroize"]
```
Comment (tony-iqlusion) [inline tai64/Cargo.toml]: Suggest removing this: the feature can be called `zeroize`

File: tai64/src/lib.rs [hunk 2]
```diff
@@ -435,3 +448,154 @@ mod tests {
         }
     }
 }
+
+// Serde Impls
```
Comment (tony-iqlusion) [inline tai64/src/lib.rs]: This crate already has `serde` support so these are unnecessary and also unrelated to `zeroize` support.

File: tai64/Cargo.toml [hunk 3]
```diff
@@ -17,10 +17,12 @@ maintenance = { status = "passively-maintained" }
 [dependencies]
 chrono = { version = "0.4", optional = true, default-features = false }
 serde = { version = "1", optional = true, default-features = false }
+zeroize = { version = "1.1.0", optional = true, default-features = false, features = ["zeroize_derive"] }
```
Comment (tony-iqlusion) [inline tai64/Cargo.toml]: If you're not using the custom derive, remove the feature for it:
```suggestion
zeroize = { version = "1", optional = true, default-features = false }
```

File: tai64/src/lib.rs [hunk 4]
```diff
@@ -40,6 +43,11 @@ const NANOS_PER_SECOND: u32 = 1_000_000_000;
 #[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
 pub struct TAI64(pub u64);
 
+#[cfg(feature = "zero_out")]
+impl Zeroize for TAI64 {
+    fn zeroize(&mut self) {}
```
Comment (tony-iqlusion) [inline tai64/src/lib.rs]: This method doesn't do anything as-is: `Zeroize` isn't just a marker trait.

You'll need to do:

```suggestion
    fn zeroize(&mut self) {
        self.0.zeroize();
    }
```

File: tai64/src/lib.rs [hunk 5]
```diff
@@ -134,6 +142,11 @@ impl Serialize for TAI64 {
 #[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
 pub struct TAI64N(pub TAI64, pub u32);
 
+#[cfg(feature = "zero_out")]
+impl Zeroize for TAI64N {
+    fn zeroize(&mut self) {}
```
Comment (tony-iqlusion) [inline tai64/src/lib.rs]: This needs to be:

```suggestion
    fn zeroize(&mut self) {
        self.0.zeroize();
        self.1.zeroize();
    }
```

File: tai64/src/lib.rs (hunk 2 above)
Comment (charleschege) [reply tai64/src/lib.rs]: I was wondering where this code for serde was coming from, I cloned the repo again and still the code seems to reappear. Turns out vscode is caching changes that I did long ago and inserting them into the new code automatically every time I open the editor. I will resolve this issue first

File: tai64/Cargo.toml (hunk 1 above)
Comment (charleschege) [reply tai64/Cargo.toml]: If I replace this with
```toml
[features]
zeroize = []
```
It results in an error `features and dependencies cannot have the same name: zeroize`

Comment (tony-iqlusion) [reply tai64/Cargo.toml]: Optional crates (somewhat confusingly) are already a cargo feature, so you can just delete this line and the `zeroize` "feature" should just work

Comment (charleschege) [reply tai64/Cargo.toml]: That works fine. Surely it's confusing

**Comment by charleschege** [comment]
All tests in the CI have passed expect the code coverage which I don't understand the issue

https://github.com/iqlusioninc/crates/runs/511913730?check_suite_focus=true

**Comment by charleschege** [comment]
Great. Can you do a minor release?

**Comment by tony-iqlusion** [comment]
Sure, will do one later today

**Pull request merged**